File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/gameroom/GameRoomCatalogSearchSupport.kt
package com.pillyliu.pinprofandroid.gameroom

import com.pillyliu.pinprofandroid.library.matchesSearchTokens
import com.pillyliu.pinprofandroid.library.normalizedSearchTokens

internal enum class GameRoomAddMachineTypeFilter(val rawValue: String, val label: String) {
    EM("em", "EM"),
    SS("ss", "SS"),
    LCD("lcd", "LCD"),
}

internal data class GameRoomCatalogSearchEntry(
    val game: GameRoomCatalogGame,
    val searchTokens: List<String>,
    val manufacturerTokens: List<String>,
)

internal fun gameRoomManufacturerSuggestions(
    options: List<String>,
    query: String,
): List<String> {
    val trimmed = query.trim()
    if (trimmed.isBlank()) return emptyList()
    val queryTokens = normalizedSearchTokens(trimmed)
    return options.filter { option ->
        matchesSearchTokens(queryTokens, normalizedSearchTokens(option))
    }.take(8)
}

internal fun buildGameRoomCatalogSearchEntries(
    games: List<GameRoomCatalogGame>,
    variantOptions: (String) -> List<String>,
): List<GameRoomCatalogSearchEntry> {
    return games.map { game ->
        val searchTokens = buildList {
            addAll(normalizedSearchTokens(game.displayTitle))
            addAll(normalizedSearchTokens(game.displayVariant.orEmpty()))
            addAll(normalizedSearchTokens(game.opdbShortname.orEmpty()))
            addAll(normalizedSearchTokens(game.opdbCommonName.orEmpty()))
            addAll(normalizedSearchTokens(game.manufacturer.orEmpty()))
            variantOptions(game.catalogGameID).forEach { variant ->
                addAll(normalizedSearchTokens(variant))
            }
        }
        GameRoomCatalogSearchEntry(
            game = game,
            searchTokens = searchTokens,
            manufacturerTokens = normalizedSearchTokens(game.manufacturer.orEmpty()),
        )
    }
}

internal fun filterGameRoomCatalogGames(
    entries: List<GameRoomCatalogSearchEntry>,
    nameQuery: String,
    manufacturerQuery: String,
    yearQuery: String,
    selectedType: GameRoomAddMachineTypeFilter?,
): List<GameRoomCatalogGame> {
    val nameTokens = normalizedSearchTokens(nameQuery.trim())
    val manufacturerTokens = normalizedSearchTokens(manufacturerQuery.trim())
    val targetYear = yearQuery.trim().toIntOrNull()
    return entries.mapNotNull { entry ->
        val game = entry.game
        val matchesName = nameTokens.isEmpty() || matchesSearchTokens(nameTokens, entry.searchTokens)
        val matchesManufacturer = manufacturerTokens.isEmpty() || matchesSearchTokens(manufacturerTokens, entry.manufacturerTokens)
        val matchesYear = targetYear == null || game.year == targetYear
        val matchesType = selectedType == null || gameRoomSearchCategory(game) == selectedType.rawValue
        if (matchesName && matchesManufacturer && matchesYear && matchesType) game else null
    }
}

internal fun gameRoomSearchCategory(game: GameRoomCatalogGame): String? {
    return when {
        game.opdbDisplay == GameRoomAddMachineTypeFilter.LCD.rawValue -> GameRoomAddMachineTypeFilter.LCD.rawValue
        game.opdbType == GameRoomAddMachineTypeFilter.EM.rawValue -> GameRoomAddMachineTypeFilter.EM.rawValue
        game.opdbType == GameRoomAddMachineTypeFilter.SS.rawValue -> GameRoomAddMachineTypeFilter.SS.rawValue
        else -> null
    }
}

File: Pinball App 2/Pinball App 2/gameroom/GameRoomCatalogSearchSupport.swift
import Foundation

enum GameRoomAddMachineTypeFilter: String, CaseIterable, Identifiable {
    case em
    case ss
    case lcd

    var id: String { rawValue }

    var label: String {
        switch self {
        case .em:
            return "EM"
        case .ss:
            return "SS"
        case .lcd:
            return "LCD"
        }
    }
}

struct GameRoomCatalogSearchEntry: Identifiable {
    let game: GameRoomCatalogGame
    let searchTokens: [String]
    let manufacturerTokens: [String]

    var id: String { game.id }
}

func gameRoomManufacturerSuggestions(options: [String], query: String) -> [String] {
    let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmed.isEmpty else { return [] }
    let queryTokens = normalizedSearchTokens(trimmed)
    return options.filter { option in
        matchesSearchTokens(queryTokens, haystackTokens: normalizedSearchTokens(option))
    }
    .prefix(8)
    .map { $0 }
}

func buildGameRoomCatalogSearchEntries(
    games: [GameRoomCatalogGame],
    variantOptions: (String) -> [String]
) -> [GameRoomCatalogSearchEntry] {
    games.map { game in
        let searchTokens = (
            [
                game.displayTitle,
                game.displayVariant,
                game.opdbShortname,
                game.opdbCommonName,
                game.manufacturer
            ]
            .compactMap { $0 } + variantOptions(game.catalogGameID)
        )
        .flatMap(normalizedSearchTokens)

        return GameRoomCatalogSearchEntry(
            game: game,
            searchTokens: searchTokens,
            manufacturerTokens: normalizedSearchTokens(game.manufacturer ?? "")
        )
    }
}

func filteredGameRoomCatalogGames(
    entries: [GameRoomCatalogSearchEntry],
    nameQuery: String,
    manufacturerQuery: String,
    yearQuery: String,
    selectedType: GameRoomAddMachineTypeFilter?
) -> [GameRoomCatalogGame] {
    let nameTokens = normalizedSearchTokens(nameQuery.trimmingCharacters(in: .whitespacesAndNewlines))
    let manufacturerTokens = normalizedSearchTokens(manufacturerQuery.trimmingCharacters(in: .whitespacesAndNewlines))
    let targetYear = Int(yearQuery.trimmingCharacters(in: .whitespacesAndNewlines))

    return entries.compactMap { entry in
        let game = entry.game
        let matchesName = nameTokens.isEmpty || matchesSearchTokens(nameTokens, haystackTokens: entry.searchTokens)
        let matchesManufacturer = manufacturerTokens.isEmpty ||
            matchesSearchTokens(manufacturerTokens, haystackTokens: entry.manufacturerTokens)
        let matchesYear = targetYear == nil || game.year == targetYear
        let matchesType = selectedType == nil || gameRoomSearchCategory(for: game) == selectedType?.rawValue
        return matchesName && matchesManufacturer && matchesYear && matchesType ? game : nil
    }
}

func gameRoomSearchCategory(for game: GameRoomCatalogGame) -> String? {
    if game.opdbDisplay == GameRoomAddMachineTypeFilter.lcd.rawValue {
        return GameRoomAddMachineTypeFilter.lcd.rawValue
    }
    if game.opdbType == GameRoomAddMachineTypeFilter.em.rawValue {
        return GameRoomAddMachineTypeFilter.em.rawValue
    }
    if game.opdbType == GameRoomAddMachineTypeFilter.ss.rawValue {
        return GameRoomAddMachineTypeFilter.ss.rawValue
    }
    return nil
}
